src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandPathExtensions.cs

```diff
@@ -0,0 +1,130 @@
+        /// For example: $expand=NS.SubType1/abc/NS.SubType2/nav
+        /// => firstPropertySegment: "abc"
+        /// => remainingSegments:  NS.SubType2/nav
+        /// => leadingTypeSegment: NS.SubType1
+        /// </summary>
+        /// <param name="expandPath">The input $expand path.</param>
+        /// <param name="remainingSegments">The remaining segments after the first non type segment.</param>
+        /// <returns>First non-type cast segment.</returns>
+        public static ODataPathSegment GetFirstNonTypeCastSegment(this ODataExpandPath expandPath,
+            out IList<ODataPathSegment> remainingSegments)
+        {
+            if (expandPath == null)
+            {
+                throw new ArgumentNullException("expandPath");
+            }
+
+            // In fact, ODataExpandPath constructor verifies the supporting segments, we add the verification here for double check.
+            return GetFirstNonTypeCastSegment(expandPath,
+                //The middle segment should be "TypeSegment" or "PropertySegment".
+                m => m is PropertySegment || m is TypeSegment,
+                // The last segment could be "NavigationPropertySegment"
+                s => s is NavigationPropertySegment,
+                out remainingSegments);
+        }
+
+        private static ODataPathSegment GetFirstNonTypeCastSegment(ODataPath path,
+            Func<ODataPathSegment, bool> middleSegmentPredicte,
+            Func<ODataPathSegment, bool> lastSegmentPredicte,
+            out IList<ODataPathSegment> remainingSegments) // could be null
+        {
+            Contract.Assert(path != null);
+
+            remainingSegments = null;
+            ODataPathSegment firstNonTypeSegment = null;
+            int lastIndex = path.Count - 1;
+            int index = 0;
+            foreach (var segment in path)
+            {
+                if (index == lastIndex)
+                {
+                    // Last segment
+                    if (!lastSegmentPredicte(segment))
+                    {
+                        throw new ODataException(Error.Format(SRResources.InvalidLastSegmentInSelectExpandPath, segment.GetType().Name));
+                    }
+                }
+                else
+                {
+                    // middle segment
+                    if (!middleSegmentPredicte(segment))
+                    {
+                        throw new ODataException(Error.Format(SRResources.InvalidSegmentInSelectExpandPath, segment.GetType().Name));
+                    }
+                }
+
+                index++;
+
+                if (firstNonTypeSegment != null)
+                {
+                    if (remainingSegments == null)
+                    {
+                        remainingSegments = new List<ODataPathSegment>();
+                    }
+
+                    Contract.Assert(remainingSegments != null);
```

Comment:
probably unnecessary given statement above

Reply:
Hmm...  Ok.

---
In reply to: [355018172](https://github.com/OData/WebApi/pull/1941#discussion_r355018172) [](ancestors = 355018172)